app/src/org/odk/collect/android/activities/FormEntryActivity.java

```diff
@@ -605,47 +607,101 @@ private void processIntentResponse(Intent response, boolean cancelled) {
         }
     }
 
-    private void updateFormRelevencies(){
+    private void updateFormRelevancies(){
         saveAnswersForCurrentScreen(DO_NOT_EVALUATE_CONSTRAINTS);
-        
+
         if(!(mCurrentView instanceof ODKView)){
-            throw new RuntimeException("Tried to update form relevency not on compound view");
+            throw new RuntimeException("Tried to update form relevancy not on compound view");
         }
-        
-        ODKView oldODKV = (ODKView)mCurrentView;
-        
-        FormEntryPrompt[] newValidPrompts = mFormController.getQuestionPrompts();
-        Set<FormEntryPrompt> used = new HashSet<>();
-        
-        ArrayList<QuestionWidget> oldWidgets = oldODKV.getWidgets();
 
-        ArrayList<Integer> removeList = new ArrayList<>();
+        ODKView odkView = (ODKView)mCurrentView;
+        ArrayList<QuestionWidget> oldWidgets = odkView.getWidgets();
+        // These 2 calls need to be made here, rather than in the for loop below, because at that
+        // point the widgets will have already started being updated to the values for the new view
+        ArrayList<Vector<SelectChoice>> oldSelectChoices = getOldSelectChoicesForEachWidget(oldWidgets);
+        ArrayList<String> oldQuestionTexts = getOldQuestionTextsForEachWidget(oldWidgets);
 
-           for(int i=0;i<oldWidgets.size();i++){
-            QuestionWidget oldWidget = oldWidgets.get(i);
-            boolean stillRelevent = false;
-
-            for(FormEntryPrompt prompt : newValidPrompts) {
-            	if(prompt.getIndex().equals(oldWidget.getPrompt().getIndex())) {
-            		stillRelevent = true;
-            		used.add(prompt);
-            	}
-            }
-            if(!stillRelevent){
-                removeList.add(i);
+        FormEntryPrompt[] newValidPrompts = mFormController.getQuestionPrompts();
```

Comment:
The issue that requires us to save the two arrays above is that we can't compare these directly (IE: "hasSameDisplayContent(FormPrompt oldFormPrompt)") because the prompt objects themselves are getting manipulated, right?